lib/Conversion/XTenNNToTosa.cpp

```diff
@@ -0,0 +1,226 @@
+//===- XTenNNToTosa.cpp -----------------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023 Advanced Micro Devices, Inc.
+//
+//===----------------------------------------------------------------------===//
+
+#include "xten/Conversion/XTenNNToTosa.h"
+#include "PassDetail.h"
+#include "xten/Dialect/XTenNN/IR/XTenNNOps.h"
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+namespace {
+
+/// Convert the quantized integer type from signed to a signless version that
+/// matches TOSA.
+///
+/// TOSA currently supports i8, i16, i32 tensors types here we convert the
+/// arbitrary signed type from the XTenNN QDQ operators to one that can be
+/// represented in TOSA.
+/// @TODO: once TOSA moves from signless to signed integers with arbitrary
+/// bit width we no longer need the type conversion.
+///
+///\param tensorType signed integer tensor type.
+///\return TensorType new storage type for the \p tensorType.
+TensorType getNewStorageType(TensorType tensorType) {
+  assert(tensorType.getElementType().isSignedInteger() &&
+         "quantization should only work with integers");
+  unsigned int integerBitwidth = tensorType.getElementTypeBitWidth();
+  if (integerBitwidth <= 8) {
+    return tensorType.cloneWith({},
+                                IntegerType::get(tensorType.getContext(), 8));
+  }
+  if (integerBitwidth <= 16) {
+    return tensorType.cloneWith({},
+                                IntegerType::get(tensorType.getContext(), 16));
+  }
+  return tensorType.cloneWith({},
+                              IntegerType::get(tensorType.getContext(), 32));
+}
+
+RankedTensorType createSplatType(int64_t rank, Type elementType) {
+  llvm::SmallVector<int64_t, 4> tmpShape;
+  for (uint32_t i = 0; i < rank; ++i) {
```

Comment:
What is the advantage? Save space when rank is != 4? What would be the default size of a small vector?

Reply:
Yes, and also you resize once to the correct size instead of resizing upon insertion when necessary. I guess it doesn't make too much of a difference here, as the rank is expected to be small.

Reply:
Ok I get your point now, I'll add the reserve real quick! Though I will keep it with a default of 4 elements. I think the default of SmallVector is 8 elements for `int64_t` and most of our tensors will be rank 4.